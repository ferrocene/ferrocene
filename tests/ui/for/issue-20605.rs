//@ revisions: current next
//@[next] compile-flags: -Znext-solver

fn changer<'a>(mut things: Box<dyn Iterator<Item=&'a mut u8>>) {
    for item in *things { *item = 0 }
    //[current]~^ ERROR `dyn Iterator<Item = &'a mut u8>` is not an iterator
    //[next]~^^ ERROR `dyn Iterator<Item = &'a mut u8>` is not an iterator
    //[next]~| ERROR the type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
    //[next]~| ERROR the type `&mut <dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
    //[next]~| ERROR the type `Option<<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item>` is not well-formed
    //[next]~| ERROR type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::Item` cannot be dereferenced

    // FIXME(-Znext-solver): these error messages are horrible and have to be
    // improved before we stabilize the new solver.
}

fn main() {}
