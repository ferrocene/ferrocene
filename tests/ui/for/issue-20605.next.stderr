error[E0277]: `dyn Iterator<Item = &'a mut u8>` is not an iterator
  --> $DIR/issue-20605.rs:6:17
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^ `dyn Iterator<Item = &'a mut u8>` is not an iterator
   |
   = help: the trait `IntoIterator` is not implemented for `dyn Iterator<Item = &'a mut u8>`

error: the type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
  --> $DIR/issue-20605.rs:6:17
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^

error: the type `&mut <dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter` is not well-formed
  --> $DIR/issue-20605.rs:6:17
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^

error: the type `Option<<<dyn Iterator<Item = &'a mut u8> as IntoIterator>::IntoIter as Iterator>::Item>` is not well-formed
  --> $DIR/issue-20605.rs:6:17
   |
LL |     for item in *things { *item = 0 }
   |                 ^^^^^^^

error[E0614]: type `<dyn Iterator<Item = &'a mut u8> as IntoIterator>::Item` cannot be dereferenced
  --> $DIR/issue-20605.rs:6:27
   |
LL |     for item in *things { *item = 0 }
   |                           ^^^^^

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0277, E0614.
For more information about an error, try `rustc --explain E0277`.
